File: common_base/src/main/java/com/xwl/common_base/dialog/UpdateDialog.java
package com.xwl.common_base.dialog;

import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.example.lib_net.download.DownloadUtil;
import com.xwl.common_base.R;
import com.xwl.common_lib.dialog.TipsToast;
import com.xwl.common_lib.utils.AppUtils;

/**
 * @author lxw
 * @date 2023/2/1
 * descripe
 */
public class UpdateDialog extends Dialog {
    private TextView tvIgnore;
    private TextView tvVersion;
    private Button btnDownload;
    private ProgressBar progressBar;
    private TextView tvProgress;
    private Context mContext;
    private ServiceConnection conn;
    private String mVersionName = "";
    private String mDownloadUrl;
    private String mFileName;
    private Intent intent;

    public UpdateDialog(@NonNull Context context, String path, String fileName) {
        super(context);
        this.mContext = context;
        this.mDownloadUrl = path;
        this.mFileName = fileName;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.update_dialog_layout);
        setCanceledOnTouchOutside(false);
        initView();
    }

    private void initView() {
        getWindow().setBackgroundDrawableResource(android.R.color.transparent);
        tvIgnore = findViewById(R.id.tv_ignore);
        btnDownload = findViewById(R.id.yes);
        tvVersion = findViewById(R.id.tv_version);
        progressBar = findViewById(R.id.progress_bar);
        tvProgress = findViewById(R.id.tv_progress);
        tvVersion.setText(mVersionName);
        btnDownload.setOnClickListener(v -> {
            btnDownload.setVisibility(View.GONE);
            progressBar.setVisibility(View.VISIBLE);
            tvProgress.setVisibility(View.VISIBLE);
            download();
        });
        tvIgnore.setOnClickListener(v -> dismiss());
    }

    private void download() {
        DownloadUtil.INSTANCE.startDownload(mContext, mDownloadUrl, mFileName, progress -> {
            progressBar.setProgress(progress);
            tvProgress.setText(String.format("%d%%", progress));
            return null;
        }, path -> {
            dismiss();
            AppUtils.installApk(mContext, path);
            return null;
        }, msg -> {
            TipsToast.INSTANCE.showTips(msg);
            return null;
        });
    }

    public void setVersionName(String versionName) {
        this.mVersionName = versionName;
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (intent != null) {
            mContext.stopService(intent);
            mContext.unbindService(conn);
            intent = null;
            conn = null;
        }

    }
}

File: mod_home/src/main/java/com/example/mod_home/dialog/UpdateDialog.kt
package com.example.mod_home.dialog

import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.FragmentActivity
import com.example.lib_net.download.DownloadUtil.startDownload
import com.example.mod_home.databinding.DialogUpdateLayoutBinding
import com.hjq.permissions.Permission
import com.sum.user.dialog.ChoosePhotoDialog
import com.xwl.common_base.dialog.BaseDialog
import com.xwl.common_base.dialog.BaseDialogFragment
import com.xwl.common_lib.callback.Permissions
import com.xwl.common_lib.dialog.TipsToast.showTips
import com.xwl.common_lib.ext.gone
import com.xwl.common_lib.ext.onClick
import com.xwl.common_lib.ext.visible
import com.xwl.common_lib.utils.AppUtils

/**
 * @author  lxw
 * @date 2023/10/19
 * descripe
 */
class UpdateDialog {
    class Builder(activity: FragmentActivity) :
        BaseDialogFragment.Builder<ChoosePhotoDialog.Builder>(activity) {
        private val mBinding: DialogUpdateLayoutBinding =
            DialogUpdateLayoutBinding.inflate(LayoutInflater.from(activity))

        /** Apk 文件  */
        private var mApkFileName: String? = null

        /** 下载地址  */
        private var mDownloadUrl: String? = null

        /** 文件 MD5  */
        private val mFileMd5: String? = null

        /** 是否强制更新  */
        private var mForceUpdate = false

        /** 当前是否下载中  */
        private val mDownloading = false

        /** 当前是否下载完毕  */
        private val mDownloadComplete = false

        init {
            initView()
        }

        private fun initView() {
            setContentView(mBinding.root)
            setHeight(ViewGroup.LayoutParams.WRAP_CONTENT)
            setAnimStyle(BaseDialog.AnimStyle.SCALE)
            gravity = Gravity.CENTER

            mBinding.tvIgnore.onClick {
                dismiss()
            }
            mBinding.updateNow.onClick {
                downloadApk()
            }
        }

        fun setVersionName(versionName: CharSequence): Builder {
            mBinding.tvVersion.text = versionName
            return this
        }

        /**
         * 设置强制更新
         */
        fun setForceUpdate(force: Boolean): Builder {
            mForceUpdate = force
            mBinding.tvIgnore.visibility = if (force) View.GONE else View.VISIBLE
            setCancelable(!force)
            return this
        }

        /**
         * 设置下载 url
         */
        fun setDownloadUrl(url: String): Builder {
            mDownloadUrl = url
            return this
        }

        @Permissions(
            Permission.WRITE_EXTERNAL_STORAGE,
            Permission.READ_MEDIA_IMAGES,
            Permission.READ_MEDIA_VIDEO,
            Permission.READ_MEDIA_AUDIO
        )
        private fun downloadApk() {
            setCancelable(false)

            mApkFileName = "mvvmarchitecture"
            // 创建要下载的文件对象
            mDownloadUrl?.let {
                mApkFileName?.let { it1 ->
                    startDownload(context, it, it1, start = {
                        mBinding.progressBar.visible()
                        mBinding.updateNow.gone()
                        mBinding.tvProgress.visible()
                    }, { progress: Int ->
                        mBinding.progressBar.progress = progress
                        mBinding.tvProgress.text = String.format("%d%%", progress)

                    }, { path: String? ->
                        dismiss()
                        AppUtils.installApk(context, path)
                    }) { msg: String? ->
                        showTips(msg)
                    }
                }
            }
        }
    }
}